### PR TITLE
Bugfix: RPC/Wallet: removeprunedfunds should flush the wallet db

### PR DESCRIPTION
**Pull request by luke-jr**
… rather than try to turn into the regular-interval wallet flushing thread

### CODE REVIEW

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -350,7 +350,8 @@ UniValue removeprunedfunds(const UniValue& params, bool fHelp)
         throw JSONRPCError(RPC_INTERNAL_ERROR, "Transaction does not exist in wallet.");
     }
 
-    ThreadFlushWalletDB(pwalletMain->strWalletFile);
+    LOCK(bitdb.cs_db);
```
Comment (MarcoFalke) [inline src/wallet/rpcdump.cpp]: Shouldn't LOCKs be scoped?

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -896,16 +910,9 @@ void ThreadFlushWalletDB(const string& strFile)
...
-                        LogPrint("db", "Flushing %s\n", strFile);
+                        FlushWalletDB(strFile);
                         nLastFlushed = nWalletDBUpdated;
-                        int64_t nStart = GetTimeMillis();
-
-                        // Flush wallet file so it's self contained
-                        bitdb.CloseDb(strFile);
-                        bitdb.CheckpointLSN(strFile);
-
                         bitdb.mapFileUseCount.erase(mi++);
```
Comment (MarcoFalke) [inline src/wallet/walletdb.cpp]: This causes the merge conflict

**Comment by MarcoFalke** [comment]
utACK 4d702d2a2d301d67a560b97d0a2b85e4d07a98c0

File: src/wallet/walletdb.cpp (hunk 2 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Does it actually conflict if you try to merge it?

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (luke-jr) [inline src/wallet/rpcdump.cpp]: Yes, but the scope here is acceptable IMO? All it does is return a const.

File: src/wallet/walletdb.cpp (hunk 2 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Nevermind, I see my local master branch was just outdated.

**Comment by luke-jr** [comment]
Rebased, no changes.

**Comment by luke-jr** [comment]
(Would appreciate @instagibbs to confirm I got the intent right here.)

**Comment by instagibbs** [comment]
@luke-jr eh, I mostly aped similar functions to be honest, I wouldn't be an expert at db flushing.

**Comment by luke-jr** [comment]
@instagibbs Well, what was the purpose you had in mind for that line? Was it intended to flush the db? (all it actually did was rename the RPC thread)

**Comment by instagibbs** [comment]
Oh, yes

**Comment by laanwj** [comment]
I don't think this is the right solution:

- The ThreadFlushDB is misnamed, it is not a flushing thread but a compaction/consolidation thread.
- There is no need to trigger a db compaction at this point. A normal `cwallet->Flush()` would do.

**Comment by luke-jr** [comment]
Depends on the goal, I guess. I assumed since this is *removing* data, compacting might be desirable to ensure it gets removed entirely.

**Comment by laanwj** [comment]
> Depends on the goal, I guess. I assumed since this is removing data, compacting might be desirable to ensure it gets removed entirely.

That's a good point in itself, however this is a very slow operation with a large wallet, so doing it after every command is non-optimal.

Also, although I called it 'compaction' it does not have any guarantee of getting rid of old data. It just consolidates the log files into `wallet.dat`. There is however a call to do a rewriting of the wallet, which will have the effect of getting rid of all unused blocks: `CDB::Rewrite`. Which we use when encrypting, but the drawback is that it requires a full restart of the client. *Maybe* adding an explicit RPC call to do this operation at some point in the future makes sense. I don't know.

But not automatically, not in this place.

**Comment by laanwj** [comment]
Closing in favor of #8765

**Pull request closed without merging**